### PR TITLE
Use `Self` where possible

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary

+ Some cleaning, can be reviewed commit per commit.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @davidhewitt

### CODE REVIEW

**Comment by Viicos** [comment]
Please review

**Review by sydney-runkle — APPROVED** [review]
Great work here!

**Comment by vikyw89** [comment]
Is this one intended to be a breaking change ?

File: pydantic/main.py [hunk 1]
```diff
@@ -1544,13 +1560,13 @@ def create_model(  # noqa: C901
...
     resolved_bases = types.resolve_bases(__base__)
-    meta, ns, kwds = types.prepare_class(__model_name, resolved_bases, kwds=__cls_kwargs__)
+    meta, ns, kwds = types.prepare_class(model_name, resolved_bases, kwds=__cls_kwargs__)
     if resolved_bases is not __base__:
         ns['__orig_bases__'] = __base__
     namespace.update(ns)
 
     return meta(
-        __model_name,
+        model_name,
```
Comment (vikyw89) [inline pydantic/main.py:1569]: this one is breaking backwards compatibility

File: pydantic/main.py [hunk 2]
```diff
@@ -1427,40 +1441,42 @@ def create_model(
     ...
 
 
-@typing.overload
+@overload
 def create_model(
-    __model_name: str,
+    model_name: str,
```
Comment (vikyw89) [inline pydantic/main.py:1446]: this one is breaking backwards compatibility

**Comment by Viicos** [comment]
@vikyw89 You can read more about it here: https://typing.readthedocs.io/en/latest/spec/historical.html#positional-only-parameters.

It is technically a breaking change, although using `__model_name` as a kw-argument would have be caught by type checkers in the first place. Sorry for the inconvenience, probably the changelog should be updated as per the issue that was opened by @fraser-langton

**Pull request merged**